Simplified/Reader2/BusinessLogic/NYPLBookmarkFactory.swift

```diff
@@ -0,0 +1,149 @@
+//
+//  NYPLBookmarkFactory.swift
+//  Simplified
+//
+//  Created by Ettore Pasquini on 3/22/21.
+//  Copyright © 2021 NYPL Labs. All rights reserved.
+//
+
+import Foundation
+import R2Shared
+
+class NYPLBookmarkFactory {
+  private let book: NYPLBook
+  private let publication: Publication
+  private let drmDeviceID: String?
+
+  init(book: NYPLBook, publication: Publication, drmDeviceID: String?) {
+    self.book = book
+    self.publication = publication
+    self.drmDeviceID = drmDeviceID
+  }
+
+  func make(from bookmarkLoc: NYPLBookmarkR2Location,
+            usingBookRegistry bookRegistry: NYPLBookRegistryProvider) -> NYPLReadiumBookmark? {
+
+    guard let progression = bookmarkLoc.locator.locations.progression else {
+      return nil
+    }
+    let chapterProgress = Float(progression)
+
+    guard let total = bookmarkLoc.locator.locations.totalProgression else {
+      return nil
+    }
+    let totalProgress = Float(total)
+
+    var page: String? = nil
+    if let position = bookmarkLoc.locator.locations.position {
+      page = "\(position)"
+    }
+
+    let registryLoc = bookRegistry.location(forIdentifier: book.identifier)
+    var cfi: String? = nil
+    var idref: String? = nil
+    if registryLoc?.locationString != nil,
+      let data = registryLoc?.locationString.data(using: .utf8),
+      let registryLocationJSON = try? JSONSerialization.jsonObject(with: data),
+      let registryLocationDict = registryLocationJSON as? [String: Any] {
+
+      cfi = registryLocationDict["contentCFI"] as? String
+
+      // backup idref from R1 in case parsing from R2 fails for some reason
+      idref = registryLocationDict["idref"] as? String
+    }
+
+    // get the idref from R2 data structures. Should be more reliable than R1's
+    // when working with R2 since it comes directly from a R2 Locator object.
+    if let idrefFromR2 = publication.idref(forHref: bookmarkLoc.locator.href) {
+      idref = idrefFromR2
+    }
+
+    let chapter: String?
+    if let locatorChapter = bookmarkLoc.locator.title {
+      chapter = locatorChapter
+    } else if let tocLink = publication.tableOfContents.first(withHREF: bookmarkLoc.locator.href) {
+      chapter = tocLink.title
+    } else {
+      chapter = nil
```

Comment:
I think we can remove `chapter = nil` here

Reply:
actually it's giving me a compile error because `chapter` is a `let` constant, so it requires explicit initialization. If it was a `var` then the `chapter = nil` statement would be redundant.

I didn't know this detail so thank you! There seem to be a proposal to make `var` work like `let` does: https://forums.swift.org/t/prepitch-optional-variables-should-require-explicit-initialization/26077